### ISSUE TITLE
add_zero_attn in MultiheadAttention breaks causality

### ISSUE THREAD

**Issue by Separius** [pytorch/pytorch#27461]
## 🐛 Bug

add_zero_attn=True in MultiheadAttention is ignoring the mask during the backward()

## To Reproduce

```python
import torch
import numpy as np

embedding_dim = 8
batch_size = 1
num_heads = 2
seq_len = 4

net = torch.nn.MultiheadAttention(embedding_dim, num_heads, add_zero_attn=True)
mask = torch.from_numpy(np.array([[True, True, True, True],
                                  [False, True, True, True],
                                  [False, False, True, True],
                                  [False, False, False, True]])).float() * -10000.0

for i in range(seq_len):
    x = torch.ones(seq_len, batch_size, embedding_dim, requires_grad=True)
    o, w = net(x, x, x, attn_mask=mask)
    # o.shape is (seq_len, batch_size, embedding_dim)
    o.mean([1, 2])[i].backward()
    print(i, x.grad.sum([1, 2]).view(-1))
```

## Expected behavior

```python
# expected output:
# 0, [0, 0, 0, 0] # it means that the first output(first token in the sequence) does not depend on any input
# 1, [x, 0, 0, 0] # it means that the second output only depends on the first token
# 2, [y, z, 0, 0]
# 3, [t, w, u, 0]

# observed output:
# 0, [0, 0, 0, 0] # this one seems right
# 1, [x, y, 0, 0] # this is obviously wrong(the provided mask should have prevented the second token from attending to itself)
# 2, [z, t, w, 0]
# 3, [u, a, b, c]
```

## Environment

```
Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Arch Linux
GCC version: (GCC) 9.2.0
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: No
CUDA runtime version: 10.1.243
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/opt/cuda/targets/x86_64-linux/lib/libcudnn.so.7.6.2
/usr/lib/libcudnn.so.7.6.4

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] numpydoc==0.9.1
[pip] pytorch-memlab==0.0.3
[pip] pytorch-transformers==1.1.0
[pip] torch==1.2.0
[pip] torchvision==0.4.0a0+6b959ee
[pip] torchviz==0.0.1
[conda] blas                      1.0                         mkl  
[conda] magma-cuda101             2.5.1                         1    pytorch
[conda] mkl                       2019.4                      243  
[conda] mkl-include               2019.4                      243  
[conda] mkl-service               2.0.2            py37h7b6447c_0  
[conda] mkl_fft                   1.0.12           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] pytorch                   1.2.0           py3.7_cuda10.0.130_cudnn7.6.2_0    pytorch
[conda] pytorch-memlab            0.0.3                    pypi_0    pypi
[conda] pytorch-transformers      1.1.0                    pypi_0    pypi
[conda] torchvision               0.4.0                py37_cu100    pytorch
[conda] torchviz                  0.0.1                    pypi_0    pypi
```

## Additional context

you can run the provided code with `add_zero_attn=False` and see that the gradient is only non zero in non masked locations (except for the full masked token)

**Comment by andreaskoepf** [comment]
I have not yet fully analyzed the problem and I am not sure if the observed effect is a bug or by design.

It seems that in the self-attention variant with the added zero k & v sequence-entries (produced internally by `add_zero_attn=True`) the gradient is composed of a component coming from the query and one from values and keys. The gradient from the query is causing the undesired effect.  

I modified the repro script a bit in order to get the effect without `add_zero_attn=True` (initial in-projection bias should be 0) and to split the contribution of query and key & value: 

```
import torch

embedding_dim = 1
batch_size = 1
num_heads = 1
seq_len = 4

net = torch.nn.MultiheadAttention(embedding_dim, num_heads, add_zero_attn=False)
mask = torch.cat([torch.ones(seq_len, seq_len).triu(), torch.zeros(seq_len, 1)], dim=1)
mask[mask==1]=float('-inf')
print(mask)

for i in range(seq_len):
    x = torch.ones(seq_len, batch_size, embedding_dim, requires_grad=True)
    y = torch.ones(seq_len, batch_size, embedding_dim, requires_grad=True)
    z = torch.cat([y, torch.zeros(1, 1, embedding_dim)])    # add zero sequence element
    o, w = net(x, z, z, attn_mask=mask)
    #print(w)
    # o.shape is (seq_len, batch_size, embedding_dim)
    o.mean([1, 2])[i].backward()
    print(i, 'x:', x.grad.abs().sum([1, 2]).view(-1))
    print(i, 'y:', y.grad.abs().sum([1, 2]).view(-1))
``` 

Output is:
```
tensor([[-inf, -inf, -inf, -inf, 0.],
        [0., -inf, -inf, -inf, 0.],
        [0., 0., -inf, -inf, 0.],
        [0., 0., 0., -inf, 0.]])
0 x: tensor([0., 0., 0., 0.])
0 y: tensor([0., 0., 0., 0.])
1 x: tensor([0.0000, 0.0148, 0.0000, 0.0000])
1 y: tensor([0.2801, 0.0000, 0.0000, 0.0000])
2 x: tensor([0.0000, 0.0000, 0.0127, 0.0000])
2 y: tensor([0.1798, 0.1798, 0.0000, 0.0000])
3 x: tensor([0.0000, 0.0000, 0.0000, 0.0105])
3 y: tensor([0.1323, 0.1323, 0.1323, 0.0000])
```

**Comment by Separius** [comment]
I think I was wrong. Based on @andreaskoepf 's experiment, it makes perfect sense for the output to be dependent on the query and there is no way to alleviate it.

Thanks.

**Comment by shisi2015** [comment]
i can't understand why add_zero_attn needed?

**Comment by netw0rkf10w** [comment]
@andreaskoepf @Separius Could you guys please explain in what context `add_zero_attn` is necessary? The only information is from the docstrings:

```
add_zero_attn: add a new batch of zeros to the key and value sequences at dim=1.
```

But why?

Thank you very much!

**Comment by Separius** [comment]
@netw0rkf10w say you want to do language modeling; One way of doing it is by adding a <start> token to the beginning of the input and set the output to the shifted input: ```input=[<start>, a, b, c], output=[a, b, c, <end_of_sequence>], attn_mask=causal```

The other way is to add a dummy token to the sequence and allow all the other tokens to attend to this dummy token so ```input=[dummy, a, b, c], output=[a, b, c], attn_mask=causal with zeroed diagonal```

Note that if you don't add this dummy token to your input, token ```a``` (in the provided example) can't look at any other token (they are all masked) and because of the way masking has been implemented (```softmax(logits-mask*10e8)```) it will instead look at "all" of the other tokens (uniformly) which is not desirable and it will break causality.

```add_zero_attn``` is an easy to use flag to do just this, so it will add a dummy zero token to the input (key and value, to be exact) and allow all of the tokens to attend to that.

One great example usage of this kind of attention is [XLNet](https://arxiv.org/abs/1906.08237), you can check Figuare1.b in the paper and see that there are tokens that can not attend to any other token, so in the implementation, you should add a dummy token to attend to.

**Comment by netw0rkf10w** [comment]
@Separius Thanks for the explanation. As I understand, the main reason is to add a dummy token so that when an output token does not have access to any of the input tokens, the attention weights corresponding to all the input tokens remain (nearly) zeros.

However, I am not totally convinced. Instead of doing

```python
attn_output_weights.masked_fill_(attn_mask, float('-inf'))
attn_output_weights = softmax(attn_output_weights, dim=-1)
```

why not do masking after computing the weights like this:

```python
attn_output_weights = softmax(attn_output_weights, dim=-1)
attn_output_weights.masked_fill_(attn_mask, 0.0)
```

or for obtaining the same weights as in the first case, now with proper masking:

```python
attn_output_weights.masked_fill_(attn_mask, float('-inf'))
attn_output_weights = softmax(attn_output_weights, dim=-1)
attn_output_weights.masked_fill_(attn_mask, 0.0)
```

If the mask is correct, then the dummy token is not needed here. Is this correct?

**Comment by Separius** [comment]
@netw0rkf10w I think your third implementation (mask, softmax, and then mask again) is totally fine and it is indeed correct but maybe (I'm not sure) it is inefficient compared to the dummy token.

I suggest you profile both methods in forward and backward propagation (and just to be safe check their gradients as well) with different batch sizes and sequence lengths and if it was indeed faster, I think it is a better solution than the current one and you can send a pull request.

@zhangguanheng66 what do you think?
